core/common/DynamicStringFormatter.cpp

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2025 iLogtail Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "common/DynamicStringFormatter.h"
+
+namespace logtail {
+
+DynamicStringFormatter::DynamicStringFormatter() = default;
+
+DynamicStringFormatter::~DynamicStringFormatter() = default;
+
+bool DynamicStringFormatter::Init(const std::string& formatString) {
+    mFormatTemplate = formatString;
+    mPlaceholders.clear();
+    mStaticParts.clear();
+
+    return ParseFormatString(formatString);
+}
+
+bool DynamicStringFormatter::Format(const ValueProvider& valueProvider, std::string& result) const {
+    if (!IsDynamic()) {
+        result = mFormatTemplate;
+        return true;
+    }
+
+    result.clear();
+
+    for (size_t i = 0; i < mPlaceholders.size(); ++i) {
+        if (i < mStaticParts.size()) {
+            result += mStaticParts[i];
+        }
+
+        const auto& placeholder = mPlaceholders[i];
+        std::string placeholderValue = valueProvider(placeholder.name);
+
+        if (placeholderValue.empty()) {
+            return false;
+        }
+
+        result += placeholderValue;
+    }
+
+    if (!mStaticParts.empty() && mPlaceholders.size() < mStaticParts.size()) {
+        result += mStaticParts.back();
+    }
+
+    return true;
+}
+
+bool DynamicStringFormatter::ParseFormatString(const std::string& formatString) {
```

Comment:
In dynamic strings, the ${} syntax semantically means to extract an environment variable and substitute its value; therefore, it should be resolved during initialization. Only %{} placeholders need to be processed dynamically at each formatting step.